### PR TITLE
GH-116017: Get rid of `_COLD_EXIT`

### PR DESCRIPTION
**Pull request by brandtbucher**
The cold exits use quite a bit of memory on JIT builds, and complicate memory allocation improvements (since they are the only immortal, shared-by-all-interpreters, unlinked, un-invalidatable executors). They also introduce a very subtle back-and-forth dance between executors when warming up side exits, and represent a significant proportion of the uops executed.

This:
- Removes them and puts all of their logic into the `_EXIT_TRACE` instruction (previously this logic was shared between the `exit_to_trace` label (only used by `_EXIT_TRACE`) and `_COLD_EXIT`...
- ...and also changes `_EXIT_TRACE` to use an `oparg` instead of an `exit_index`...
- ...which allows us to simplify the tier two instruction format.

Separately, I added some missing stats to `_DYNAMIC_EXIT`.

This results in a 34.7% reduction in traces executed, and a 1.6% reduction in uops executed (both corresponding to [3.7 billion fewer `_COLD_EXITS`](https://github.com/faster-cpython/benchmarking-public/blob/main/results/bm-20240618-3.14.0a0-a1bdb94-JIT/bm-20240618-azure-x86_64-brandtbucher-no_cold_exits-3.14.0a0-a1bdb94-pystats-vs-base.md)). Per platform:
- `aarch64-apple-darwin`: [3% faster](https://github.com/faster-cpython/benchmarking-public/blob/main/results/bm-20240618-3.14.0a0-a1bdb94-JIT/bm-20240618-darwin-arm64-brandtbucher-no_cold_exits-3.14.0a0-a1bdb94-vs-base.svg), [9% less memory](https://github.com/faster-cpython/benchmarking-public/blob/main/results/bm-20240618-3.14.0a0-a1bdb94-JIT/bm-20240618-darwin-arm64-brandtbucher-no_cold_exits-3.14.0a0-a1bdb94-vs-base-mem.svg)
- `aarch64-unknown-linux-gnu`: [9% faster](https://github.com/faster-cpython/benchmarking-public/blob/main/results/bm-20240618-3.14.0a0-a1bdb94-JIT/bm-20240618-arminc-aarch64-brandtbucher-no_cold_exits-3.14.0a0-a1bdb94-vs-base.svg), [2% less memory](https://github.com/faster-cpython/benchmarking-public/blob/main/results/bm-20240618-3.14.0a0-a1bdb94-JIT/bm-20240618-arminc-aarch64-brandtbucher-no_cold_exits-3.14.0a0-a1bdb94-vs-base-mem.svg)
- `i686-pc-windows-msvc`: [3% faster](https://github.com/faster-cpython/benchmarking-public/blob/main/results/bm-20240618-3.14.0a0-a1bdb94-JIT/bm-20240618-pythonperf1_win32-x86-brandtbucher-no_cold_exits-3.14.0a0-a1bdb94-vs-base.svg)
- `x86_64-pc-windows-msvc`: [3% faster](https://github.com/faster-cpython/benchmarking-public/blob/main/results/bm-20240618-3.14.0a0-a1bdb94-JIT/bm-20240618-pythonperf1-amd64-brandtbucher-no_cold_exits-3.14.0a0-a1bdb94-vs-base.svg)
- `x86_64-unknown-linux-gnu`: [1% faster](https://github.com/faster-cpython/benchmarking-public/blob/main/results/bm-20240618-3.14.0a0-a1bdb94-JIT/bm-20240618-linux-x86_64-brandtbucher-no_cold_exits-3.14.0a0-a1bdb94-vs-base.svg), [2% less memory](https://github.com/faster-cpython/benchmarking-public/blob/main/results/bm-20240618-3.14.0a0-a1bdb94-JIT/bm-20240618-linux-x86_64-brandtbucher-no_cold_exits-3.14.0a0-a1bdb94-vs-base-mem.svg)

* Issue: gh-116017

### CODE REVIEW

**Comment by Fidget-Spinner** [comment]
Maybe a question for @mdboom instead of you: why are the figures on the time plot different than the figures given in geometric mean? E.g. darwin arm64 says "Geometric mean: 1.00x faster" but the figure shows 1.027x" or roughly 3%?

**Comment by brandtbucher** [comment]
I think the number in the README overview of the results is the "HPT" calculation, which (I believe) weights by "noise". You can see that those particular results were pulled upwards by the startup benchmarks, which have a higher standard deviation than most of the others.

But yeah, @mdboom can probably explain better.

**Comment by mdboom** [comment]
The plot takes the mean across all of the distributions of all of the benchmarks. So it takes into account that the distributions of all of the benchmarks aren't necessarily normally distributed.

The geometric mean value (from pyperf) is the nth root of multiplying all of the means of all the benchmarks together.

File: Python/bytecodes.c [hunk 1]
```diff
@@ -4213,50 +4256,24 @@ dummy_func(
...
-            * one to keep the executor alive when executing. */
-            Py_INCREF(executor);
-            exit->executor = executor;
-            GOTO_TIER_TWO(executor);
-        }
-
         tier2 op(_DYNAMIC_EXIT, (--)) {
             tstate->previous_executor = (PyObject *)current_executor;
             _PyExitData *exit = (_PyExitData *)&current_executor->exits[oparg];
+            PyCodeObject *code = _PyFrame_GetCode(frame);
```
Comment (markshannon) [inline Python/bytecodes.c]: This is unused outside of the debug code below, so move it down 2 lines.

**Review by markshannon — COMMENTED** [review]
One minor nit, otherwise it looks good.

File: Python/bytecodes.c (hunk 1 above)
Comment (brandtbucher) [reply Python/bytecodes.c]: It's used when `target->op.code == ENTER_EXECUTOR` below.

Comment (markshannon) [reply Python/bytecodes.c]: So it is. I didn't see that you'd removed the later definition.

I'd still rather keep this off the fast path, though.

**Pull request merged**